components/support/interrupt/src/scopes.rs

```diff
@@ -0,0 +1,90 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::{Interrupted, Interruptee};
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+// Shared counter for InterruptScope:
+//   - The `interrupt()` method increments this.
+//   - The `was_interrupted()` method checks if this was incremented since the `InterruptScope` was
+//     created.
+static COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+/// Scope for an interruptable operations
+#[derive(Clone, Debug)]
+pub struct InterruptScope {
+    start_value: usize,
+}
+
+/// Scope for interruptible operations
+///
+/// This struct provides interruption support for syncing and other long-running operations.  The
+/// basic system is:
+///
+///    - Call `InterruptScope::new()` to create a new `InterruptScope` at the begining of the
+///      operation.  If multiple components are involved in the same operation, then they should
+///      share clones of a single `InterruptScope`.  For example `SyncManager.sync()` creates an
+///      interrupt scope that all `SyncEngine` implementations share.
+///    - During the operation, regularly check if the scope is interrupted with `was_interrupted()`
+///      or `err_if_interrupted()` which is often more ergonomic.
+///    - Call `InterruptScope::interrupt_current_scopes()` to interrupt all previously created
+///      `InterruptScope`s. `InterruptScope`s created after this call will not be considerid
+///      interrupted unless `InterruptScope::interrupt()` is called again.
+///
+/// This type requires the code to actively if it's interrupted.  Therefore:
+///   - Make sure to sprinkle in `err_if_interrupted()` calls inside the code you want
+///     interruptable.  Loops are a particularly good place to put these.
+///   - This cannot interrupt external code.  In particular, it won't interrupt a long-running SQLite
+///     query.  Use an `rusqlite::InterruptHandler` for that.
```

Comment:
So this is what I don't quite understand in this patch - how does interrupting a sync also arrange to interrupt any running sql, other than that special places support?

Reply:
Long-term, I'd love to make more components work like places:  Have a connection dedicated to syncing, so we can always interrupt it.  I guess the other option would be to ignore the issue.  If the other components don't have long-running queries, maybe that's an option.